ingest/expression_files/dense_ingestor.py

```diff
@@ -0,0 +1,187 @@
+an dense matrix.
+
+"""
+import collections
+import csv
+import datetime
+import sys
+from typing import List  # noqa: F401
+
+from bson.objectid import ObjectId
+
+try:
+    from expression_files import GeneExpression
+    sys.path.append("../ingest")
+    from ingest_files import IngestFiles
+
+except ImportError:
+    # Used when importing as external package, e.g. imports in single_cell_portal code
+    from .expression_files import GeneExpression
+    sys.path.append("../ingest")
+    from ..ingest_files import IngestFiles
+
+
+class DenseIngestor(GeneExpression, IngestFiles):
+    ALLOWED_FILE_TYPES = ["text/csv",
+                          "text/plain", "text/tab-separated-values"]
+
+    def __init__(self, file_path, study_file_id, study_id, **kwargs):
+        GeneExpression.__init__(self, file_path, study_file_id, study_id)
+        IngestFiles.__init__(
+            self, file_path, allowed_file_types=self.ALLOWED_FILE_TYPES
+        )
+        self.matrix_params = kwargs
+        self.csv_file_handler, self.file_handler = self.open_file(
+            self.file_path)
+        self.gene_names = {}
+        self.header = next(self.csv_file_handler)
+
+    def execute_ingest(self):
+        # Row after header is needed for R format validation
+        row = next(self.csv_file_handler)
+        if DenseIngestor.is_valid_format(self.header, row):
+            # Reset csv reader to first gene row
+            self.csv_file_handler = self.open_file(self.file_path)[0]
+            next(self.csv_file_handler)
+            for gene_docs, data_array_documents in self.transform():
+                self.load(gene_docs, data_array_documents)
+        else:
+            raise ValueError("Dense matrix has invalid format")
+
+    def matches_file_type(file_type):
+        return file_type == 'dense'
+
+    @staticmethod
+    def format_gene_name(gene):
+        return gene.strip().strip('\"')
+
+    @staticmethod
+    def process_row(row: str):
+        """
+        Performs pre-processing steps for a single row that converts gene scores
+        into floats.
+        """
+        def convert_to_float(value: str):
+            # Remove white spaces and quotes
+            value = value.strip("'\",")
+            # Convert string to float and round 3 places
+            return round(float(value), 3)
+        result = map(convert_to_float, row[1:])
+        return list(result)
+
+    @staticmethod
+    def filter_expression_scores(scores: List, cells: List):
+        """
+        Filters non-zero gene scores and corresponding cell names
+
+        Returns:
+            tuple (generator):
+                valid_expression_scores (list): non-zero gene scores
+                associated_cells (list): cells
+        """
+        associated_cells = []
+        valid_expression_scores = []
+        for idx, expression_score in enumerate(scores, 1):
+            if expression_score != 0:
```

Comment:
In anticipation of aggregated cell matrices (where "genes not assayed in one of the combined studies" get NaN values to indicate "not assayed") we should also filter out NaN as they can be treated as zero for visualization.

Reply:
4402ffd48b30f870f1f2edb04f3db548748a3b13

Reply:
Per [Devon's comment](https://github.com/broadinstitute/scp-ingest-pipeline/pull/108#issuecomment-666375511) filtering out NaN is something we do now in production. My comment was just indicating that there will be more NaN in future so it is best that we maintain parity with production now.

Reply:
Happy to see the new NaN handling!